### ISSUE TITLE
@switch type narrowing doesn't work

### ISSUE THREAD

**Issue by ArielGueta** [angular/angular#52077]
### Which @angular/* package(s) are the source of the bug?

compiler

### Is this a regression?

No

### Description

```ts
interface Base {
  type: 'foo' | 'bar'
}

interface Foo extends Base {
  type: 'foo';
  foo: string;
}

interface Bar extends Base {
  type: 'bar';
  bar: number;
}

@Component({
  selector: 'app-root',
  standalone: true,
  template: `
@switch (v.type) {
  @case ('bar') {
    {{ v.bar }}
  }

  @case ('foo') {
    {{ v.foo }}
  }

  @default {
    Default
  }
}
  `
})
export class AppComponent {
   v: Foo | Bar = { type: 'foo', foo: 'foo' };
}
```

<img width="580" alt="Screenshot 2023-10-06 at 16 52 13" src="https://github.com/angular/angular/assets/20522870/e5f8f688-38e9-4c6b-afb4-dae8048ab656">

### Anything else?

It should work similar to typescript code.

**Comment by JoostK** [comment]
The reason this doesn't work is because the TCB contains a parenthesized expression as switch-condition, which is not accounted for in TS's CFA.

```ts
function _tcb1(this: AppComponent) { if (true) {
    switch (((((this).v /*10,11*/) /*10,11*/).type /*12,16*/) /*10,16*/) {
        case "bar" /*29,34*/:
            "" + (((((this).v /*45,46*/) /*45,46*/).bar /*47,50*/) /*45,50*/);
            break;
        case "foo" /*68,73*/:
            "" + (((((this).v /*84,85*/) /*84,85*/).foo /*86,89*/) /*84,89*/);
            break;
        default: break;
    }
} }
```

**Comment by ArielGueta** [comment]
It should be fixed, right?

**Comment by crisbeto** [comment]
It might be worth changing switch to use if statements in the TCB as well.

**Comment by crisbeto** [comment]
Confirmed Joost's explanation here: https://www.typescriptlang.org/play?#code/JYOwLgpgTgZghgYwgAgEJwM4oN4ChnJgCeADhAFzIDkMA9rVcgD7UBGcUVuAvrrqJFiIUAMXrIIAD0ggAJhjSYc+QqQrU6DANwrNlDGCigA5jt79w0eEkVQJ0iHIXosyPAWJlKVdpx0FfShAAVwBbVmgzPhhgkAQwYFoQZGMIMAA1AAoASkoxWmZbNxUCKDTgqGSQgBtqgEIo3AQkg2QAN2QAXhS0rOydXAB6QeQAdVooAGsMXAwAd2AwBAALZEy2gDpPCGziggQlNg4qchLkZpAMWmqIDeraY3WN336z1jK4SYH9w5p6E7OFyuNzuDyemleBACHy+PD4w2QABFaBAFCBaGBkHMJpNZgslqtMk9ttldu5zr9fACoecWtdbvdHpsXv4oe8IJ9vhTXH8GKcaUD6aCmRsIazoRzYdwgA

**Comment by JeanMeche** [comment]
FWIW the root issue is fixed on TS' side microsoft/TypeScript#56035

**Comment by crisbeto** [comment]
I think we'll still have to keep the workaround around until we drop support for TS 5.2.
